Don't exit tablet server on reloading invalid ACL

wiebeytec opened the pull request:
This fixes potentially bringing down a tablet with an innocuous SIGHUP.

It also logs the fact it's reading the ACL file, to fix not getting any feedback on SIGHUP.

https://github.com/vitessio/vitess/issues/17139

## Description

Currently, deploying new ACL files is not only very involved because I have to test them against a running Vitess instance, but it's also very stressful to do `killall -SIGHUP vttablet` knowing that it can bring down the entire system.

Because of the risk of downtime caused by simple ACL definition mistakes, I think this is a candidate for back-porting.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/17139

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

dbussink commented on go/vt/tableacl/tableacl.go:
```diff
@@ -105,6 +105,7 @@ func (tacl *tableACL) init(configFile string, aclCB func()) error {
 	if configFile == "" {
 		return nil
 	}
+	log.Infof("Reading ACL file %v", configFile)
```
I don't think we should log it here. What's the reason you need it here? Was it not clear that this was happening? I would imagine it was clear from the lines below on errors what the cause of the problem was?

Below here, there's also `Info` logging on errors which is wrong, those should be raised at least to warning if not error. It's confusing those are set at the info level. 

If there's a good reason to add the logging, maybe instead we should log inside the SIGHUP handler to make it specific for the scenario? I don't think this line is something useful to print on each startup then. We should be careful with not adding too much logging noise that loses signal then.

wiebeytec replied:
> I don't think we should log it here. What's the reason you need it here? Was it not clear that this was happening? I would imagine it was clear from the lines below on errors what the cause of the problem was?

The issue is that when there is no error, sending a SIGHUP doesn't cause any log output to be generated. So, you don't know if it even did anything.

I considered adding the line to the SIGHUP handler. We can do that too I suppose. I'll change it, along with the sign-off.

dbussink replied:
> The issue is that when there is no error, sending a SIGHUP doesn't cause any log output to be generated. So, you don't know if it even did anything.

Let's log it then after it succeeds, so that we only have the line if there was a successful reload due to a `SIGHUP`. We can then make it more something like "Successfully reloaded ACL".

dbussink replied:
> you have no idea what it was trying to do.

I'm not sure what you mean here? This is something that logging can do properly? Looking at the code more I think it would benefit from some additional refactor and remove the exit at a distance kinda behavior and use error returns and let the callers decide.

wiebeytec replied:
> I'm not sure what you mean here?

I think we're miscommunicating a bit. You said:

> What's the reason you need it here? Was it not clear that this was happening? I would imagine it was clear from the lines below on errors what the cause of the problem was?

Indeed, on error. I meant that as a sysop, I would like to see a program respond to something. If I send it a SIGHUP and no log lines are printed, I can't tell whether it did something or not. 

> Looking at the code more I think it would benefit from some additional refactor and remove the exit at a distance kinda behavior and use error returns and let the callers decide.

OK, I can do that.

dbussink replied:
@wiebeytec https://github.com/wiebeytec/vitess/compare/main...dbussink:vitess:acl-refactor-suggestion?expand=1 is kinda what I was thinking about here. Dunno what you think of that?

wiebeytec replied:
Looks good. In the branch I just force-pushed I still logged in the callees (deep), but functionality it's similar.

You can remove my code in favor of yours then.

dbussink replied:
I think logging at the higher level would be preferable here.

dbussink reviewed (commented):
You'll need to amend your commit with the proper sign off to pass the DCO check as well.

wiebeytec commented:
What's the next step here? I think we can go with your code.

dbussink commented:
> What's the next step here? I think we can go with your code.

We can update your PR here with that code? Feel free to use it!

wiebeytec commented:
Alright, I updated the PR and added a test case that tests loading an invalid ACL does not remove the existing one.

mattlord commented on go/cmd/vttablet/cli/cli.go:
```diff
@@ -248,7 +248,10 @@ func createTabletServer(ctx context.Context, env *vtenv.Environment, config *tab
 	})
 	servenv.OnClose(qsc.StopService)
-	qsc.InitACL(tableACLConfig, enforceTableACLConfig, tableACLConfigReloadInterval)
+	err := qsc.InitACL(tableACLConfig, tableACLConfigReloadInterval)
+	if err != nil && enforceTableACLConfig {
+		return nil, fmt.Errorf("failed to initialize table acl: %w", err)
+	}
```
This is not about reloading the file. We should fail to start when given an invalid config and IMO it's quite dangerous to just log an error and continue on as you'd likely have far more permissive permissions than you intended/expected (and how closely are you watching tablet logs when there's otherwise no visible user impacting issues?).

mattlord replied:
Ah, sorry! We ARE returning an error here. Reading is hard 😆

mattlord reviewed (approved):
This makes sense to me. I'm leaning towards not backporting it for two reasons as I'm not confident in how this is used across the Vitess user base:
  1. What does this mean if the operator isn't watching the logs? Couldn't this potentially mean that the vttablet is running along as if everything is fine while having more permissive permissions than intended/expected? We can say that this is the operators concern/fault, but it's still a change.
  2. Some people could be relying on the current behavior (I've made this mistake in the past)

It's also for this reason that I think we should add a small blurb about this here https://github.com/vitessio/vitess/blob/main/changelog/22.0/22.0.0/changelog.md

What do others think?

wiebeytec commented:
People who are relying on that behavior would have to have a weird setup, because they need to have a running tablet server that they expect to exit after SIGHUP when the file contains errors. One example use is deployment on the replicas as a test, but then they'd be putting their replicas at risk; especially because it calls the OS `exit` system call, so it's an unclean shutdown (probably?).

Having said that, I do see your point that it's a change.

deepthi commented:
Can you add a note to the release notes as Manan requested? That will need to go into https://github.com/vitessio/vitess/blob/main/changelog/22.0/22.0.0/summary.md. You can put it under Minor Changes.

wiebeytec commented:
Sorry, I missed that was for me :grimacing:

I added it to the summary.md. I also rebased and force-pushed the branch on origin/main, to include the recent changes.

wiebeytec commented:
The CI is blocking it because the 'needs release notes' label is attached.

https://github.com/vitessio/vitess/actions/runs/13069154230/job/36705441216

I think I don't have rights to remove the label?

mattlord commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -155,6 +156,10 @@ While the flag will continue to accept float values (interpreted as seconds) for
 
+#### <a id="reloading-vttablet-acl"/>VTTablet ACL enforcement and reloading</a>
+
+When a tablet is started with `--enforce-tableacl-config`, the enforcement by exiting is only done on start, not on reload when receiving a SIGHUP. When reloading an invalid ACL, it's logged as error and the active ACL remains unchanged.
```
```suggestion
When a tablet is started with `--enforce-tableacl-config` it will exit with an error if the contents of the file are not valid. After the changes made in https://github.com/vitessio/vitess/pull/17485 the tablet will no longer exit when reloading the contents of the file after receiving a SIGHUP. When the file contents are invalid on reload the tablet will now log an error and the active in-memory ACLs remain in effect.
```
